crates/bevy_asset/src/assets.rs

```diff
@@ -1,572 +1,501 @@
+        self.allocator.recycle(index);
+        value
     }
 
-    /// Get mutable access to the asset for the given handle.
-    ///
-    /// This is the main method for mutably accessing asset data from an [Assets] collection. If you
-    /// do not need mutable access to the asset, you may also use [get](Assets::get).
-    pub fn get_mut(&mut self, handle: &Handle<T>) -> Option<&mut T> {
-        let id: HandleId = handle.into();
-        self.events.send(AssetEvent::Modified {
-            handle: Handle::weak(id),
-        });
-        self.assets.get_mut(&id)
+    pub(crate) fn get(&self, index: AssetIndex) -> Option<&A> {
+        let entry = self.storage.get(index.index as usize)?;
+        match entry {
+            Entry::None => None,
+            Entry::Some { value, generation } => {
+                if *generation == index.generation {
+                    value.as_ref()
+                } else {
+                    None
+                }
+            }
+        }
     }
 
-    /// Gets a _Strong_ handle pointing to the same asset as the given one.
-    pub fn get_handle<H: Into<HandleId>>(&self, handle: H) -> Handle<T> {
-        Handle::strong(handle.into(), self.ref_change_sender.clone())
+    pub(crate) fn get_mut(&mut self, index: AssetIndex) -> Option<&mut A> {
+        let entry = self.storage.get_mut(index.index as usize)?;
+        match entry {
+            Entry::None => None,
+            Entry::Some { value, generation } => {
+                if *generation == index.generation {
+                    value.as_mut()
+                } else {
+                    None
+                }
+            }
+        }
     }
 
-    /// Gets mutable access to an asset for the given handle, inserting a new value if none exists.
-    ///
-    /// # Events
-    ///
-    /// * [`AssetEvent::Created`]: Sent if the asset did not yet exist with the given handle.
-    pub fn get_or_insert_with<H: Into<HandleId>>(
-        &mut self,
-        handle: H,
-        insert_fn: impl FnOnce() -> T,
-    ) -> &mut T {
-        let mut event = None;
-        let id: HandleId = handle.into();
-        let borrowed = self.assets.entry(id).or_insert_with(|| {
-            event = Some(AssetEvent::Created {
-                handle: Handle::weak(id),
-            });
-            insert_fn()
+    pub(crate) fn flush(&mut self) {
+        let new_len = self
+            .allocator
+            .next_index
+            .load(std::sync::atomic::Ordering::Relaxed);
+        self.storage.resize_with(new_len as usize, || Entry::Some {
+            value: None,
+            generation: 0,
         });
-
-        if let Some(event) = event {
-            self.events.send(event);
+        while let Ok(recycled) = self.allocator.recycled_receiver.try_recv() {
+            let entry = &mut self.storage[recycled.index as usize];
+            *entry = Entry::Some {
+                value: None,
+                generation: recycled.generation,
+            };
```

Comment:
I'm not sure I see the robustness argument for the `while let` <-> `let new_len` swap. From my perspective it makes it less robust because it would then needlessly fail in a case where we recycle an id that hasn't been flushed/inserted yet (not currently possible, but we might in the future). This is totally fine provided it is monotonically increasing (which it is). I do like the idea to add the comment though.